File: 1101-1200/1182-Shortest Distance to Target Color/1182-Shortest Distance to Target Color.cpp
class Solution {
public:
    vector<int> shortestDistanceColor(vector<int>& colors, vector<vector<int>>& queries) {
        int n = colors.size();
        vector<vector<int>> dist(n, vector<int>(3, -1));
        dist[0][colors[0] - 1] = 0;
        for (int i = 1; i < n; ++i) {
            for (int c = 0; c < 3; ++c) {
                if (dist[i - 1][c] != -1) {
                    dist[i][c] = dist[i - 1][c] + 1;
                }
            }
            dist[i][colors[i] - 1] = 0;
        }
        
        for (int i = n - 2; i >= 0; --i) {
            for (int c = 0; c < 3; ++c) {
                if (dist[i + 1][c] != -1 && (dist[i][c] == -1 || dist[i + 1][c] + 1 < dist[i][c])) {
                    dist[i][c] = dist[i + 1][c] + 1;
                }
            }
        }
        
        vector<int> result(queries.size());
        for (int i = 0; i < queries.size(); ++i) {
            result[i] = dist[queries[i][0]][queries[i][1] - 1];
        }
        
        return result;
    }
};


class Solution2 {
public:
    vector<int> shortestDistanceColor(vector<int>& colors, vector<vector<int>>& queries) {
        vector<vector<int>> table(4);
        for (int i = 0; i < colors.size(); ++i) {
            table[colors[i]].push_back(i);
        }
        
        vector<int> result;
        for (auto& q : queries) {
            int i = q[0], c = q[1];
            if (table[c].size() > 0) {
                auto it = lower_bound(table[c].begin(), table[c].end(), i);
                if (it == table[c].begin()) {
                    result.push_back(*it - i);
                }
                else if (it == table[c].end()) {
                    result.push_back(i - table[c].back());
                }
                else {
                    int a = *it - i;
                    int b = i - *(--it);
                    result.push_back(min(a, b));
                }
            }
            else {
                result.push_back(-1);
            }
        }
        
        return result;
    }
};

File: 1101-1200/1182-Shortest Distance to Target Color/1182-Shortest Distance to Target Color.py
import collections
import bisect
class Solution:
    def shortestDistanceColor(self, colors: List[int], queries: List[List[int]]) -> List[int]:
        table = collections.defaultdict(list)
        for i, c in enumerate(colors):
            table[c].append(i)
            
        result = []
        for i, c in queries:
            if c in table:
                index = bisect.bisect_left(table[c], i)
                if index == 0:
                    result.append(table[c][0] - i)
                elif index == len(table[c]):
                    result.append(i - table[c][-1])
                else:
                    result.append(min(table[c][index] - i, i - table[c][index - 1]))
            else:
                result.append(-1)
        return result


class Solution2:
    def shortestDistanceColor(self, colors: List[int], queries: List[List[int]]) -> List[int]:
        dist = [[-1] * 3 for _ in range(len(colors))]
        dist[0][colors[0] - 1] = 0
        for i in range(1, len(colors)):
            for c in range(3):
                if dist[i - 1][c] != -1:
                    dist[i][c] = dist[i - 1][c] + 1
            dist[i][colors[i] - 1] = 0

        for i in range(len(colors) - 2, -1, -1):
            for c in range(3):
                if dist[i + 1][c] != -1 and (dist[i][c] == -1 or dist[i + 1][c] + 1 < dist[i][c]):
                    dist[i][c] = dist[i + 1][c] + 1
                    
        return [dist[i][c - 1] for i, c in queries]
